### ISSUE TITLE
Experiment to find the root cause of the bootlenecks

### ISSUE THREAD

**Issue by echarles** [jupyterlab/benchmarks#49]
Experiment to find the root cause of the bootlenecks in https://github.com/jupyterlab/benchmarks/tree/master/experiments

**Comment by echarles** [comment]
We had a meeting with @ellisonbg @jasongrout and @goanpeca to review the codemirror issues and the resize behavior in jlab/lumino. The following scenario tries to describe the current behavior to better understand what is going on based on the simple experiment of https://github.com/jupyterlab/benchmarks/tree/ee123a929d7130b8e9148ad70efcf15d0462b296/experiments/jlab-cell-boxpanel (1000 jlab cells wrapped into a lumino boxpanel)

From the profiling, we see the repetition on each cell add of the following patttern with force layout which bring additional latency compared to 1000 jlab cells not wrapped into a lumino boxpanel)

<img width="391" alt="0" src="https://user-images.githubusercontent.com/226720/98918618-1fc81000-24ce-11eb-948b-906ec3f34bdb.png">

We the inspect the methods that generate a force layout. The root is at the lumino `processMessage` method which is triggered by the following chain (the very first source is an update-request event)

<img width="232" alt="1" src="https://user-images.githubusercontent.com/226720/98918799-5b62da00-24ce-11eb-8c03-09f78178c8cf.png">

We zoom at the `processMessage` method

<img width="472" alt="2" src="https://user-images.githubusercontent.com/226720/98918870-70d80400-24ce-11eb-9cd0-f911aed55f7f.png">

and then at the `onResize`  located in `@jupyterlab/codemirror/lib/widget.js`

<img width="538" alt="3" src="https://user-images.githubusercontent.com/226720/98918901-7c2b2f80-24ce-11eb-988d-84186622c9cd.png">

We go then into `@jupyterlab/codemirror/lib/editor.js` where ` refresh` is invoked that finally triggers the CodeMirror refresh method that will end up in a forced layout.

<img width="405" alt="4" src="https://user-images.githubusercontent.com/226720/98919242-dd530300-24ce-11eb-9454-4dafa6100e9c.png">

**Comment by echarles** [comment]
So prolly something could be done on jlab cells level rather than lumino.

@ellisonbg has suggested to test with simple DIVs in a Lumino BoxPanel, which has been done in https://github.com/jupyterlab/benchmarks/tree/b9695a8d8b948e01dc56e43a6099fe8767ba0709/experiments/lumino-div-boxpanel

That experiment does not show any latency when adding div, which tends to confirm the above outcomes: The forced refresh is triggered by jlab cells.

<img width="793" alt="Screenshot 2020-11-12 at 10 13 16" src="https://user-images.githubusercontent.com/226720/98920227-23f52d00-24d0-11eb-9a88-43c6c511effd.png">

**Comment by echarles** [comment]
A bit more experiments on the jlab side. 

1. I have disabled in `codeeditor/src/widget.ts`

```js
   protected onUpdateRequest(msg: Message): void {
+    /*
     if (this.isVisible) {
       this._hasRefreshedSinceAttach = true;
+      this.editor.refresh();  
     }
+    */
   }
```

and in `codemirror/src/widget.ts`

   protected onUpdateRequest(msg: Message): void {
+    /*
     if (this.isVisible) {
       this._hasRefreshedSinceAttach = true;
     }
+    */

This gives without user impact an improvement from 15 seconds to 10 seconds to load a 50 codecells notebook, so 5 seconds faster, with the following pattern

<img width="635" alt="Screenshot 2020-11-12 at 13 41 51" src="https://user-images.githubusercontent.com/226720/98942882-f9ff3300-24ee-11eb-93ef-af2a97f55490.png">

You can see 3 parts: the fromJson, the setOption, and the third part which was before much larger due to forcedLayout. So we are good on part 2, let's see about part 2 which gives forcedLayout in every setOption.

We have many places in `codemirror/src/editor.ts` where cm.setOption is called. I have removed them, and we go from 10 second to 5 seconds - In that case, the cells need to be clicked to show the content.

So

- Removing the call to refresh: 15s -> 10s without impact on user side.
- Disabling the setOption: 10s -> 5s - I have looked at the codemirror api but there is no way to run batch options setting which could help

**Comment by goanpeca** [comment]
>Disabling the setOption: 10s -> 5s - I have looked at the codemirror api but there is no way to run batch options setting which could help

@echarles would this be an opportunity to wrap all the `setOption` inside an `operation` ?

**Comment by echarles** [comment]
> @echarles would this be an opportunity to wrap all the setOption inside an operation ?

That is a great idea. I was struggling with 2 issues with all those setOption:

1. There is no CodeMirror API to batch the setting of the options => You idea solve it.
2. The request to set options is spread all across the `codemirror/src/editor.ts` and I am not sure if there is an easy way to concentrate them at one place, not sure either is a method is exposed publically and can be call by an external component to set an option. If we can not make it perfect, we could at least concentrate as much as possible in the editor.ts and use the CodeMirror operation to barch multiple option settings.

**Comment by goanpeca** [comment]
> If we can not make it perfect, we could at least concentrate as much as possible in the editor.ts and use the CodeMirror operation to barch multiple option settings.

Yes, that is what I thought, so concentrate as much as possible in a single place. Maybe create a new `setOptions` method and try to use that instead (internally this method would use an operation).

**Comment by goanpeca** [comment]
>not sure either is a method is exposed publically and can be call by an external component to set an option.

Well, if there is the method could be updated to take an `{}`?to set multiple options at once also?

**Comment by goanpeca** [comment]
So we need maybe to turn this

```
  export function setOption<K extends keyof CodeMirrorEditor.IConfig>(
    editor: CodeMirror.Editor,
    option: K,
    value: CodeMirrorEditor.IConfig[K],
    config: CodeMirrorEditor.IConfig
  ):
```

Into a `setOptions`

**Comment by echarles** [comment]
> Into something like setOptions and try to remove as much of the setOption calls as possible (and centralize them)

Thx @goanpeca Would be great to see your how you can progress with this. Maybe @jasongrout can give advice and background on this?
